### ISSUE TITLE
Update to kBuild based on make 4.x

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/kbuild-os2#5]
This was actually done on branch `master` a while ago but it led to numerous problems, as described here http://trac.netlabs.org/kbuild/ticket/136 so we had to stick with the last revision of kmk prior to make 4 switch (see `master-pre-make4` branch, which is now the default one.

Now we updated `master` branch to the latest revision r3583 (b364c37fc0e8a2316ece53707eefea98b5a0ae6f, 144ea623de9f7a508f4210aeb585a9f69d71777c) but it appears that some of the problems remain.

When updating, we also picked the following changes from `master-pre-make4` and applied them at `master`:

* 39fbc27f835d463c8a16885c232a76d39d27783b and d197e375ed672dbe2c78f55313ffe0d826e7a484 both applied as 6be1fd5de5c1285e2e3cca4d4bd36e272aabc03e
* 07e5a5c8bb41e2a9270024bed125446a51c77275 applied as edeee89aca08256aa4e2bd92009fc7a6e0e570f4
* e037f1b15486020487907b1cf9e09589217bcf12 applied as 00dd2c9e1a04de725c7559c0ea96c28c4fba2c52 

At least two problems are still there:

1. Hundreds of empty xxxxxxx.tmp that remain not deleted (e.g. ~460 of them are created when building LIBCx with new kmk).
2. Lots of `fcntl(): Invalid argument` messages in kmk output.

Note that the LIBCx build works per se. I.e. I get all the DLLs and EXEs built successfully and all tests pass. I.e at least the crash mentioned in ticket 136 seems to have gone now. For this reason, 81697aa9925fcd86a2285d308939b8442c8b0647 is not applied. Seems like it's not necessary any more. And all other patches from `master-pre-make4` are on `master`.

We need to fix the above two issues, give it some more testing. If it proves to work fine, then we will make `master` branch default again.

**Comment by dmik** [comment]
It's important to note that the above problems are only seen with `kmk`. With `kmk_gmake` (a "canonical" build of GNU Make, named `make.exe` in our RPMS) there are no `fcntl` messages or `.tmp` files. All seems to run smoothly.

**Comment by dmik** [comment]
The issue with `.tmp` files and `fcntl` errors is more or less clear now. In make 4 there is a new feature called [output synchronisation during parallel execution](https://www.gnu.org/software/make/manual/make.html#Parallel-Output) and controlled by -O option. In canonical GNU `make`, no synchronisation is done by default (equivalent to `-Onone`). However, `kmk` sets it to `-Otarget` by default. Another difference between `make` and `kmk` is that the former always runs one job by default unless `-jN` is explicitly set while `kmk` runs in parallel mode by default. Given that output synchronisation only takes place when there is more than one job and something other than `-Onone` is given, both parallel mode and output synchronisation need to be enabled for `make` in order to make its this new code path.

Going further, it appears that output synchronisation fails to work properly on OS/2 because POSIX `tmpfile` implementation  in kLIBC is incomplete and because `fcntl(F_SETLKW)` is not supported on STDOUT/STDERR handles in LIBCx (and perhaps in kLIBC either). I will check LIBCn/LIBCx and create the corresponding tickets there. For kmk we will just stick with no output synchronisation on OS/2.

Perhaps it makes sense to disable this path in make completely given that both kLIBC/LIBCx don't work properly in this regard for now.

**Comment by dmik** [comment]
BTW, it's also interesting that in make 4.4 they dropped usage of `tmpfile` and `fcntl(F_SETLKW)` from the output synchronisation implementation completely (perhaps because this wasn't a very efficient solution platform-wide in the first place). However, a quick look at make 4.4 shows that they rely on anonymous temporary files there which isn't truly supported on OS/2 either, but that's another story.

**Comment by dmik** [comment]
With the above commit the task appears to be done. I will create a new task for -O option as a reminder.
